File: MPChartLib/src/com/github/mikephil/charting/highlight/Highlight.java

package com.github.mikephil.charting.highlight;

/**
 * Contains information needed to determine the highlighted value.
 * 
 * @author Philipp Jahoda
 */
public class Highlight {

    /** the x-index of the highlighted value */
    private int mXIndex;

    /** the y-value of the highlighted value */
    private float mValue = Float.NaN;

    /** the index of the data object - in case it refers to more than one */
    private int mDataIndex;

    /** the index of the dataset the highlighted value is in */
    private int mDataSetIndex;

    /** index which value of a stacked bar entry is highlighted, default -1 */
    private int mStackIndex = -1;

    /** the range of the bar that is selected (only for stacked-barchart) */
    private Range mRange;

    /**
     * constructor
     *
     * @param x the index of the highlighted value on the x-axis
     * @param value the y-value of the highlighted value
     * @param dataIndex the index of the Data the highlighted value belongs to
     * @param dataSetIndex the index of the DataSet the highlighted value belongs to
     */
    public Highlight(int x, float value, int dataIndex, int dataSetIndex) {
        this.mXIndex = x;
        this.mValue = value;
        this.mDataIndex = dataIndex;
        this.mDataSetIndex = dataSetIndex;
    }
    /**
     * Constructor, only used for stacked-barchart.
     * 
     * @param x the index of the highlighted value on the x-axis
     * @param value the y-value of the highlighted value
     * @param dataIndex the index of the Data the highlighted value belongs to
     * @param dataSetIndex the index of the DataSet the highlighted value belongs to
     * @param stackIndex references which value of a stacked-bar entry has been
     *            selected
     */
    public Highlight(int x, float value, int dataIndex, int dataSetIndex, int stackIndex) {
        this(x, value, dataIndex, dataSetIndex);
        mStackIndex = stackIndex;
    }

    /**
     * Constructor, only used for stacked-barchart.
     *
     * @param x the index of the highlighted value on the x-axis
     * @param value the y-value of the highlighted value
     * @param dataIndex the index of the Data the highlighted value belongs to
     * @param dataSetIndex the index of the DataSet the highlighted value belongs to
     * @param stackIndex references which value of a stacked-bar entry has been
     *            selected
     * @param range the range the selected stack-value is in
     */
    public Highlight(int x, float value, int dataIndex, int dataSetIndex, int stackIndex, Range range) {
        this(x, value, dataIndex, dataSetIndex, stackIndex);
        this.mRange = range;
    }

    /**
     * Constructor, only used for stacked-barchart.
     *
     * @param x the index of the highlighted value on the x-axis
     * @param dataSetIndex the index of the DataSet the highlighted value belongs to
     */
    public Highlight(int x, int dataSetIndex) {
        this(x, Float.NaN, 0, dataSetIndex, -1);
    }

    /**
     * returns the index of the highlighted value on the x-axis
     *
     * @return
     */
    public int getXIndex() {
        return mXIndex;
    }

    /**
     * returns the y-value of the highlighted value
     *
     * @return
     */
    public float getValue() {
        return mValue;
    }

    /**
     * the index of the data object - in case it refers to more than one
     *
     * @return
     */
    public int getDataIndex() {
        return mDataIndex;
    }

    /**
     * returns the index of the DataSet the highlighted value is in
     *
     * @return
     */
    public int getDataSetIndex() {
        return mDataSetIndex;
    }

    /**
     * Only needed if a stacked-barchart entry was highlighted. References the
     * selected value within the stacked-entry.
     * 
     * @return
     */
    public int getStackIndex() {
        return mStackIndex;
    }

    /**
     * Returns the range of values the selected value of a stacked bar is in. (this is only relevant for stacked-barchart)
     * @return
     */
    public Range getRange() {
        return mRange;
    }

    /**
     * returns true if this highlight object is equal to the other (compares
     * xIndex and dataSetIndex)
     * 
     * @param h
     * @return
     */
    public boolean equalTo(Highlight h) {

        if (h == null)
            return false;
        else {
            if (this.mDataSetIndex == h.mDataSetIndex && this.mXIndex == h.mXIndex
                    && this.mStackIndex == h.mStackIndex)
                return true;
            else
                return false;
        }
    }

    @Override
    public String toString() {
        return "Highlight, xIndex: " + mXIndex + ", dataSetIndex: " + mDataSetIndex
                + ", stackIndex (only stacked barentry): " + mStackIndex;
    }
}

File: MPChartLib/src/main/java/com/github/mikephil/charting/highlight/Highlight.kt
package com.github.mikephil.charting.highlight

import com.github.mikephil.charting.components.YAxis.AxisDependency
import java.io.Serializable

/**
 * Contains information needed to determine the highlighted value.
 */
class Highlight : Serializable {
    /**
     * the x-value of the highlighted value
     */
    var x: Float = Float.NaN
        private set

    /**
     * the y-value of the highlighted value
     */
    var y: Float = Float.NaN
        private set

    /**
     * the x-pixel of the highlight
     */
    var xPx: Float = 0f
        private set

    /**
     * the y-pixel of the highlight
     */
    var yPx: Float = 0f
        private set

    /**
     * the index of the data object - in case it refers to more than one
     */
    var dataIndex: Int = -1

    /**
     * the index of the dataset the highlighted value is in
     */
    val dataSetIndex: Int

    /**
     * index which value of a stacked bar entry is highlighted, default -1
     */
    var stackIndex: Int = -1
        private set

    /**
     * the axis the highlighted value belongs to
     */
    var axis: AxisDependency? = null
        private set

    /**
     * the x-position (pixels) on which this highlight object was last drawn
     */
    var drawX: Float = 0f
        private set

    /**
     * the y-position (pixels) on which this highlight object was last drawn
     */
    var drawY: Float = 0f
        private set

    constructor(x: Float, y: Float, dataSetIndex: Int, dataIndex: Int) {
        this.x = x
        this.y = y
        this.dataSetIndex = dataSetIndex
        this.dataIndex = dataIndex
    }

    constructor(x: Float, y: Float, dataSetIndex: Int) {
        this.x = x
        this.y = y
        this.dataSetIndex = dataSetIndex
        this.dataIndex = -1
    }

    constructor(x: Float, dataSetIndex: Int, stackIndex: Int) : this(x, Float.NaN, dataSetIndex) {
        this.stackIndex = stackIndex
    }

    /**
     * constructor
     *
     * @param x            the x-value of the highlighted value
     * @param y            the y-value of the highlighted value
     * @param dataSetIndex the index of the DataSet the highlighted value belongs to
     */
    constructor(x: Float, y: Float, xPx: Float, yPx: Float, dataSetIndex: Int, axis: AxisDependency?) {
        this.x = x
        this.y = y
        this.xPx = xPx
        this.yPx = yPx
        this.dataSetIndex = dataSetIndex
        this.axis = axis
    }

    /**
     * Constructor, only used for stacked-barchart.
     *
     * @param x            the index of the highlighted value on the x-axis
     * @param y            the y-value of the highlighted value
     * @param dataSetIndex the index of the DataSet the highlighted value belongs to
     * @param stackIndex   references which value of a stacked-bar entry has been
     * selected
     */
    constructor(x: Float, y: Float, xPx: Float, yPx: Float, dataSetIndex: Int, stackIndex: Int, axis: AxisDependency?) : this(
        x,
        y,
        xPx,
        yPx,
        dataSetIndex,
        axis
    ) {
        this.stackIndex = stackIndex
    }

    val isStacked: Boolean
        get() = stackIndex >= 0

    /**
     * Sets the x- and y-position (pixels) where this highlight was last drawn.
     *
     * @param x
     * @param y
     */
    fun setDraw(x: Float, y: Float) {
        this.drawX = x
        this.drawY = y
    }

    /**
     * Returns true if this highlight object is equal to the other (compares
     * xIndex and dataSetIndex)
     *
     * @param h
     * @return
     */
    fun equalTo(h: Highlight?): Boolean {
        return if (h == null)
            false
        else
            this.dataSetIndex == h.dataSetIndex && this.x == h.x && this.stackIndex == h.stackIndex && this.dataIndex == h.dataIndex
    }

    override fun toString(): String {
        return "Highlight, x:$x y:$y dataSetIndex:$dataSetIndex stackIndex (only stacked barentry): $stackIndex"
    }
}
